src/monodroid/jni/embedded-assemblies.cc

```diff
@@ -125,114 +126,248 @@ EmbeddedAssemblies::TypeMappingInfo_compare_key (const void *a, const void *b)
+		for (; idx < module_count; idx++) {
+			const uint8_t *java_entry = binary_search<const char, uint8_t, compare_java_name, true> (java_type_name.get (), modules[idx].java_map, modules[idx].entry_count, modules[idx].java_name_width + 3);
+			if (java_entry == nullptr)
+				continue;
+			type_token_id = *reinterpret_cast<const int32_t*>(java_entry + modules[idx].java_name_width);
+			break;
+		}
+
+		if (idx >= module_count) {
+			log_error (LOG_ASSEMBLY, "typemap: unable to find module with Java type '%s' mapping", java_type_name.get ());
+			return nullptr;
+		}
+
+		module = &modules[idx];
+	} else {
 #endif
-	return find_entry_in_type_map (java, jm_typemap, jm_typemap_header);
-}
+		const TypeMapJava *java_entry = binary_search<const char, TypeMapJava, compare_java_name, true> (java_type_name.get (), map_java, java_type_count, java_name_width);
+		if (java_entry == nullptr) {
+			log_warn (LOG_ASSEMBLY, "typemap: unable to find mapping to a managed type from Java type '%s'", java_type_name.get ());
+			return nullptr;
+		}
 
-const char*
-EmbeddedAssemblies::typemap_managed_to_java (const char *managed)
-{
+		if (java_entry->module_index >= map_module_count) {
+			log_warn (LOG_ASSEMBLY, "typemap: mapping from Java type '%s' to managed type has invalid module index", java_type_name.get ());
+			return nullptr;
+		}
+
+		module = const_cast<TypeMapModule*>(&map_modules[java_entry->module_index]);
+		const TypeMapModuleEntry *entry = binary_search <int32_t, TypeMapModuleEntry, compare_type_token> (&java_entry->type_token_id, module->map, module->entry_count);
+		if (entry == nullptr) {
+			log_warn (LOG_ASSEMBLY, "typemap: unable to find mapping from Java type '%s' to managed type with token ID %u in module [%s]", java_type_name.get (), java_entry->type_token_id, mono_guid_to_string (module->module_uuid));
+			return nullptr;
+		}
+		type_token_id = java_entry->type_token_id;
 #if defined (DEBUG) || !defined (ANDROID)
-	for (TypeMappingInfo *info = managed_to_java_maps; info != nullptr; info = info->next) {
-		/* log_warn (LOG_DEFAULT, "# jonp: checking file: %s!%s for type '%s'", info->source_apk, info->source_entry, managed); */
-		const char *e = reinterpret_cast <const char*> (bsearch (managed, info->mapping, static_cast<size_t>(info->num_entries), static_cast<size_t>(info->entry_length), TypeMappingInfo_compare_key));
-		if (e == nullptr)
-			continue;
-		return e + info->value_offset;
 	}
 #endif
-	return find_entry_in_type_map (managed, mj_typemap, mj_typemap_header);
-}
 
-#if defined (DEBUG) || !defined (ANDROID)
-void
-EmbeddedAssemblies::extract_int (const char **header, const char *source_apk, const char *source_entry, const char *key_name, int *value)
-{
-	int    read              = 0;
-	int    consumed          = 0;
-	size_t key_name_len      = 0;
-	char   scanf_format [20] = { 0, };
+	if (module->image == nullptr) {
+		module->image = mono_image_loaded (module->assembly_name);
+		if (module->image == nullptr) {
+			// TODO: load
```

Comment:
This TODO is troubling.

Consider e.g. 807e66501257d2adee731df1ed501301e2083530 / PR #4058: an e.g. `Activity` can come from *any* assembly within the `.apk`; it need not be present "only" within the "main" assembly.

Nor, for that matter, is there anything which ensures or requires that an Activity type be in an already loaded assembly, especially since (before this PR) we'd use `Type.GetType()` to load the Activity type, which would implicitly load the assembly, as part of `JNIEnv.RegisterJniNatives()`.

I suspect that this TODO will in fact break that: if you have an app w/ a "Main Launcher" Activity located in a separate assembly, and you launch the app via that icon, things will break.

Reply:
That's why it is still a TODO, I need to implement the load logic. However, note that we **still** do `Type.GetType()` from `JNIENV.RegisterJniNatives()` - so that invariant remains and hence the TODO is not that dangerous or urgent to fix. I'm yet to see it fail in my tests (which doesn't mean it doesn't have to be implemented)

Reply:
By any chance is there an open issue related to this TODO?  If not, we should likely add one, so that the commit message can properly reference it.